src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -116,11 +123,66 @@ public Http2Connection(HttpConnectionPool pool, Stream stream)
             _maxConcurrentStreams = int.MaxValue;
             _pendingWindowUpdate = 0;
 
+            if (_pool.EnableMultipleHttp2Connections)
+            {
+                _freeStreamSlots = InitialSlotsLimit;
+            }
+
             if (NetEventSource.Log.IsEnabled()) TraceConnection(stream);
         }
 
         private object SyncObject => _httpStreams;
 
+        public bool AcquireStreamSlot()
+        {
+            int currentSlots = Volatile.Read(ref _freeStreamSlots);
+
+            while (true)
+            {
+                if (currentSlots == 0)
+                {
+                    long currentExpiration = Interlocked.Read(ref _freeStreamSlotsExpiration);
```

Comment:
All of this attempt at lock-freedom makes me nervous; I'd bet money there's a reliability bug lurking here, just given the complexity.  Is it all necessary?

Reply:
This method was simplified down to a regular CAS and repeat loop.